Sources/eppo/EppoClient.swift

```diff
@@ -97,147 +130,100 @@ public class EppoClient {
+        
+        if let variation = flagEvaluation.variation, !isValueOfType(expectedType: expectedVariationType, variationValue: variation.value) {
+            throw Errors.variationWrongType
         }
         
         // Optionally log assignment
-        if let assignmentLogger = self.assignmentLogger {
-            let assignment = Assignment(
-                flagKey: flagKey,
-                allocationKey: rule.allocationKey,
-                variation: assignedVariation.value,
-                subject: subjectKey,
-                timestamp: ISO8601DateFormatter().string(from: Date()),
-                subjectAttributes: subjectAttributes
-            )
-            
-            // Prepare the assignment cache key
-            let assignmentCacheKey = AssignmentCacheKey(
-                subjectKey: subjectKey,
-                flagKey: flagKey,
-                allocationKey: rule.allocationKey,
-                variationValue: assignedVariation.typedValue
-            )
-            
-            // Check if the assignment has already been logged, if the cache is defined
-            if let cache = self.assignmentCache, cache.hasLoggedAssignment(key: assignmentCacheKey) {
-                // The assignment has already been logged, do nothing
-            } else {
-                // Either the cache is not defined, or the assignment hasn't been logged yet
-                assignmentLogger(assignment)
-                self.assignmentCache?.setLastLoggedAssignment(key: assignmentCacheKey)
+        if flagEvaluation.doLog {
+            if let allocationKey = flagEvaluation.allocationKey,
+               let variation = flagEvaluation.variation,
+               let assignmentLogger = self.assignmentLogger {
+                
+                // Prepare the assignment cache key
+                let assignmentCacheKey = AssignmentCacheKey(
+                    subjectKey: subjectKey,
+                    flagKey: flagKey,
+                    allocationKey: allocationKey,
+                    variationValue: variation.value
+                )
+                
+                // Check if the assignment has already been logged, if the cache is defined
+                if let cache = self.assignmentCache, cache.hasLoggedAssignment(key: assignmentCacheKey) {
+                    // The assignment has already been logged, do nothing
+                } else {
+                    // Either the cache is not defined, or the assignment hasn't been logged yet
+                    // Perform assignment.
+                    let assignment = Assignment(
+                        flagKey: flagKey,
+                        allocationKey: allocationKey,
+                        variation: variation.key,
+                        subject: subjectKey,
+                        timestamp: ISO8601DateFormatter().string(from: Date()),
+                        subjectAttributes: subjectAttributes
+                    )
```

Comment:
The UFC has two new logging fields, `extra logging` and `metadata,` which we should bake into here as well.